### PR TITLE
3003 review entity validation

### LINKED ISSUE #3003: Ensure full backend entity metadata validation
Backend needs to ensure metadata consistency:

- [x] - should not allow incorrect foreign keys (either because they are deleted entries or because they belong to different templates or thesauris)

- [x] -  should not allow incorrect types
- [x] -  should not allow 'rogue' property names not found on the template
- [ ] -  check if ti merits to do this for batch processes as well.
   - [x] - saveMultiple on our entitiesModel (this calls directly model.save, not executing validation) i think we need to add validation but this will result in slower saves, and we do not really know how many entitites are we saving at a time using this function.
   - [ ] - updateMany using mongo query like { $set/$rename/$unset : {} }, i do not know how we can validate this properly as you can send things like {'metadata.name': 'name'}, maybe check how on how many cases we using this and change it so that it is always (if possible) an entity structure that we can validate using the current validator.

### PR DESCRIPTION
**Pull request by daneryl**
fixes #3003 

PR checklist:
- [ ] Update READ.me ?
- [ ] Update API documentation ?

QA checklist:
- [ ] Smoke test the functionality described in the issue
- [ ] Test for side effects
- [ ] UI responsiveness
- [ ] Cross browser testing
- [ ] Code review

### CODE REVIEW

**Comment by fnocetti** [comment]
@daneryl, great job!
@mfacar and I did the QA and found the following:

- foreingIds check is implemented for relationship fields but not for invalid thesauri on select/multiselect fields. Was that left behind on purpose?
- When doing a multi-update, the check for properties not declared in the template is not working. That is, excecuting a multiupdate on a property "nonexistent" that doesn't exist in the template of the entity is persisting that field to mongo.
- As for the rest of the validations, they are working like charm but we are seeing some weird validation messages on return. Consider this case (i'm pasting an example curl request without the cookie):

```
curl 'http://localhost:3000/api/entities' \
  -H 'Connection: keep-alive' \
  -H 'Accept: application/json' \
  -H 'X-Requested-With: XMLHttpRequest' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.106 Safari/537.36' \
  -H 'Content-Language: en' \
  -H 'Content-Type: application/json' \
  -H 'Origin: http://localhost:3000' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Dest: empty' \
  -H $'Referer: http://localhost:3000/en/uploads/?q=(allAggregations:\u0021f,includeUnpublished:\u0021f,order:desc,sort:creationDate,unpublished:\u0021t)' \
  -H 'Accept-Language: en-US,en;q=0.9' \
  --data-binary '{"metadata":{"text":[{"value":"texto"}],"numeric":[{"value":4}],"select":[{"value":""}],"multi_select":[],"date_range":[{"value":{"from":1593648000,"to":1593820799}}],"multi_date":[{"value":1594252800},{"value":1594771200}],"multi_date_range":[{"value":{"from":1595808000,"to":1596239999}},{"value":{"from":1595980800,"to":1596239999}}],"rich_text":[{"value":"as"}],"image":[{"value":""}],"preview":[{"value":""}],"media":[{"value":""}],"link":[{"value":{"label":1,"url":"www.google.com"}}],"geolocation_geolocation":[{"value":{"lat":45.42176635207197,"lon":4.813476562500087,"label":""}}]},"template":"5efe117f1cc91aad46472075","title":"eB","type":"entity"}' \
  --compressed
```

In that case the label for the link field is (wrongly) a number. The validation is succeeding to stop the request, but the message shows as follows:

```
{
   "error":"validation failed",
   "validations":[
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/0/type",
         "params":{
            "type":"null"
         },
         "message":"should be null"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/1/type",
         "params":{
            "type":"string"
         },
         "message":"should be string"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/2/type",
         "params":{
            "type":"number"
         },
         "message":"should be number"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value.label",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/3/properties/label/oneOf/0/type",
         "params":{
            "type":"string"
         },
         "message":"should be string"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value.label",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/3/properties/label/oneOf/1/type",
         "params":{
            "type":"null"
         },
         "message":"should be null"
      },
      {
         "keyword":"oneOf",
         "dataPath":".metadata['link'][0].value.label",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/3/properties/label/oneOf",
         "params":{
            "passingSchemas":null
         },
         "message":"should match exactly one schema in oneOf"
      },
      {
         "keyword":"additionalProperties",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/4/additionalProperties",
         "params":{
            "additionalProperty":"label"
         },
         "message":"should NOT have additional properties"
      },
      {
         "keyword":"additionalProperties",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/4/additionalProperties",
         "params":{
            "additionalProperty":"url"
         },
         "message":"should NOT have additional properties"
      },
      {
         "keyword":"additionalProperties",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/5/additionalProperties",
         "params":{
            "additionalProperty":"url"
         },
         "message":"should NOT have additional properties"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value.label",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/5/properties/label/type",
         "params":{
            "type":"string"
         },
         "message":"should be string"
      },
      {
         "keyword":"required",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/5/required",
         "params":{
            "missingProperty":"lat"
         },
         "message":"should have required property 'lat'"
      },
      {
         "keyword":"required",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/5/required",
         "params":{
            "missingProperty":"lon"
         },
         "message":"should have required property 'lon'"
      },
      {
         "keyword":"type",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf/6/type",
         "params":{
            "type":"array"
         },
         "message":"should be array"
      },
      {
         "keyword":"oneOf",
         "dataPath":".metadata['link'][0].value",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf/0/items/properties/value/oneOf",
         "params":{
            "passingSchemas":null
         },
         "message":"should match exactly one schema in oneOf"
      },
      {
         "keyword":"anyOf",
         "dataPath":".metadata['link']",
         "schemaPath":"#/properties/metadata/additionalProperties/anyOf",
         "params":{

         },
         "message":"should match some schema in anyOf"
      },
      {
         "message":"should be a \"{ label: string, url: string }\" object properties can not be blank",
         "dataPath":".metadata['link']",
         "schemaPath":"#/metadataMatchesTemplateProperties"
      }
   ],
   "prettyMessage":"validation failed"
}
```

We are not sure if this behavior happens in other cases too.
All of the tests were carried through the API.

**Comment by daneryl** [comment]
@fnocetti thanks for the review, working on the changes 👍

**Comment by daneryl** [comment]
> foreingIds check is implemented for relationship fields but not for invalid thesauri on select/multiselect fields. Was that left behind on purpose?

not left behind on pupose, forgot about it 🙈, its implemented now.

**Comment by daneryl** [comment]
> As for the rest of the validations, they are working like charm but we are seeing some weird validation messages on return. Consider this case (i'm pasting an example curl request without the cookie):

all this errors is the result of using nested oneOf key on our metadata "static" validation, ajv reports all the possible values this can contain, even on nested oneOf keys, i think is really confusing also, i would not include a "fix" for this as part of this PR but definitely we want better error reports, by post processing them or using something like https://atlassian.github.io/better-ajv-errors/

**Comment by fnocetti** [comment]
> > As for the rest of the validations, they are working like charm but we are seeing some weird validation messages on return. Consider this case (i'm pasting an example curl request without the cookie):
> 
> all this errors are the result of using nested oneOf key on our metadata "static" validation, ajv reports all the possible values this can contain, even on nested oneOf keys, i think is really confusing also, i would not include a "fix" for this as part of this PR but definitely we want better error reports, by post processing them or using something like https://atlassian.github.io/better-ajv-errors/

Yeah, completely agree! Thanks for the research!

**Comment by fnocetti** [comment]
> > foreingIds check is implemented for relationship fields but not for invalid thesauri on select/multiselect fields. Was that left behind on purpose?
> 
> not left behind on pupose, forgot about it 🙈, its implemented now.
> 
> > When doing a multi-update, the check for properties not declared in the template is not working. That is, excecuting a multiupdate on a property "nonexistent" that doesn't exist in the template of the entity is persisting that field to mongo.
> 
> i can not replicate this locally, pls can you provide me with a step by step scenario ?

We re-checked the scenario and couldn't reproduce neither. Sorry about that, we must have confused the tests.

**Comment by fnocetti** [comment]
@daneryl the unit tests show some problems with typings. Please, take a look.
I reviewed the code for the latest changes and LGTM.
We will QA the thesauri validation when the UT are green :)

**Comment by daneryl** [comment]
@fnocetti fixed the tests, there was actually a bug on thesauri validation regarding values with subgroups, it is fixed now.

File: app/shared/types/validateMetadataField.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    ).map(v => v.id);
+
+    const diff = value
+      .filter(v => v.value)
+      .filter(v => !dictionaryValues.includes(String(v.value)));
+
+    if (diff.length) {
+      return [
+        validationError(
+          { message: customErrorMessages.relationship_wrong_foreign_id, data: diff },
```
Comment (mfacar) [inline app/shared/types/validateMetadataField.ts:83]: @daneryl  could this message be more related to the validation of the dictionary? Getting "related entities do not exist or belong to another template" perhaps is a little confusing.

**Comment by mfacar** [comment]
> @fnocetti fixed the tests, there was actually a bug on thesauri validation regarding values with subgroups, it is fixed now.

Thanks @daneryl, all works as expected. In order to merge the PR, please review the comment about the message of thesauri validation.

File: app/shared/types/validateMetadataField.ts (hunk 1 above)
Comment (daneryl) [reply app/shared/types/validateMetadataField.ts]: @mfacar thanks, i have made it so relationship and dictionary have different errors 👍

**Pull request merged**